src/Microsoft.Performance.SDK.Runtime/Extensibility/CustomDataProcessorExtensibilitySupport.cs

```diff
@@ -37,35 +39,86 @@ public class CustomDataProcessorExtensibilitySupport
             this.tableReferences = new Dictionary<TableDescriptor, ITableExtensionReference>();
         }
 
         /// <inheritdoc />
-        public bool AddTable(TableDescriptor tableDescriptor)
+        public void EnableTable(TableDescriptor tableDescriptor)
         {
             lock (this.dataProcessor)
             {
-                if (this.dataExtensionRetrievalFactory != null)
+                if (this.finalizedData)
                 {
-                    throw new InvalidOperationException($"Cannot add tables after calling {nameof(FinalizeTables)}.");
+                    throw new InvalidOperationException(
+                        $"Tables cannot be enabled after finalizing the {nameof(IDataProcessorExtensibilitySupport)}.");
                 }
             }
 
-            if (tableDescriptor.RequiredDataCookers.Any() || tableDescriptor.RequiredDataProcessors.Any())
+            if (!tableDescriptor.RequiresDataExtensions())
+            {
+                return;
+            }
+
+            if (!TableExtensionReference.TryCreateReference(tableDescriptor.Type, true, out var tableReference))
+            {
+                // todo: create TableExtensionReference.CreateReference (throws) to get a more descriptive exception
+                throw new ExtensionTableException(
+                    $"Unable to create {nameof(ITableExtensionReference)} for {tableDescriptor.Type}");
+            }
+
+            if (tableReference.Availability == DataExtensionAvailability.Undetermined)
             {
-                if (TableExtensionReference.TryCreateReference(tableDescriptor.Type, true, out var tableReference))
+                tableReference.ProcessDependencies(this.dataExtensionRepository);
+            }
+
+            if (tableReference.Availability != DataExtensionAvailability.Available)
+            {
+                throw new ExtensionTableException($"The table is not available: {tableReference.Availability}");
+            }
+
+            if (!tableReference.IsInternalTable)
+            {
+                EnableRequiredSourceDataCookers(tableReference);
+                return;
+            }
+
+            // check that there are no external sources required.
+            foreach (var cookerPath in tableReference.DependencyReferences.RequiredSourceDataCookerPaths)
+            {
+                if (!StringComparer.Ordinal.Equals(cookerPath.SourceParserId, this.dataProcessor.SourceParserId))
                 {
-                    this.tableReferences[tableDescriptor] = tableReference;
-                    return true;
+                    throw new ExtensionTableException(
+                        $"The internal table {tableReference.TableDescriptor.Type} references source parser " +
+                        $"{cookerPath.SourceParserId} and cannot be enabled on {this.dataProcessor.SourceParserId}.");
                 }
             }
 
+            this.tableReferences[tableDescriptor] = tableReference;
```

Comment:
I noticed we have a lock earlier in the methods, which suggests that this may be called from multiple threads. Should we be protecting this update here?